### DISCUSSION TITLE
How to Pass File Paths as Parameters for Batch Processing?

### DISCUSSION THREAD

**Discussion by hasibagen** [quarto-dev/quarto-cli#10081] [Q&A] [answered]
### Description

Issue Description:

I would like to ask how to pass the file path as a parameter to achieve batch processing of large amounts of data. This is my current code:

```
quarto render test_par.ipynb --to docx -P sample_path="/home/_2024_05_10_15_44_16.csv"
pandoc
```
The code seems to run successfully without any errors, but it appears that the parameter is not being passed correctly.

```
to: docx
output-file: test_par.docx
default-image-extension: png
ascii: false

metadata
title: test_par
mainfont: STZhongsong
monofont: STZhongsong
papersize: a4

Output created: test_par.docx
```

Here is the relevant portion of my Quarto file and the reproducible demo file:
```
---
title: "test_par"
mainfont: "STZhongsong"
monofont: "STZhongsong"
format:  docx
params:
  sample_path: 
execute:
  enabled: true
  echo: false
  warning: false
  eval: true
ascii: false
fig-dpi: 300
papersize: a4
---

#| tags: [parameters]
print(sample_path)

```
[test_par.zip](https://github.com/user-attachments/files/15921887/test_par.zip)

**Comment by hasibagen** [comment]
I have referred to
https://github.com/quarto-dev/quarto-cli/discussions/9107
and added
execute:
enabled: true

**Comment by hasibagen** [comment]
Is this completely impossible?

It seems that this post mentioned that jupter does not support parameter passing?
https://github.com/quarto-dev/quarto-cli/discussions/5941

But I may prefer to use jupter at present.

**Comment by mcanouil** [comment]
Please make sure your code is properly formatted and is correct.
You are missing the code cell, you have a YAML key (params) without a value.

Also be sure to read the documentation: https://quarto.org/docs/computations/parameters.html#jupyter


The options `papersize` or the font options does not apply to `docx` format, please be sure to read the guide and reference pages of the documentation:
- https://quarto.org/docs/reference/formats/docx.html
- https://quarto.org/docs/output-formats/ms-word.html
---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Comment by hasibagen** [comment]
My question is how to allow ipynb to also pass in parameters directly

minimal self-contained reproducible Quarto document is here
qmd file:

``````
---
title: "Reproducible Quarto Document"
format:  docx
params:
  datapath: null
execute:
  enable : true
  echo: false
  warning: false
  eval: true
jupyter: python3
---

# markdown

```{python}
#| tags: [parameters]
#
```
```{python}
#| label: import
#| tags: [parameters]
print(datapath)
``````

jupyter ipynb is

```
---
title: "Reproducible Quarto Document"
format:  docx
params:
  samplepath: null
engine: jupyter
---
#| tags: [parameters]
print(samplepath)
# markdown
```
[bug_rep.zip](https://github.com/user-attachments/files/15923841/bug_rep.zip)

Run the code and the output is

qmd

```
quarto render bug_rep.qmd -P datapath:1

Executing 'bug_rep.quarto_ipynb'
  Cell 1/3: ''.........Done
  Cell 2/3: ''.........Done
  Cell 3/3: 'import'...Done

pandoc
  to: docx
  output-file: bug_rep.docx
  default-image-extension: png

metadata
  title: Reproducible Quarto Document
  jupyter: python3

Output created: bug_rep.docx
Reproducible Quarto Document
1
markdown

```
![截图 2024-06-21 14-45-08](https://github.com/quarto-dev/quarto-cli/assets/12797356/2609770c-5cbf-4d2c-9b23-de3dd119e355)


ipynb
```
quarto render bug_rep.ipynb -P datapath:1
pandoc
  to: docx
  output-file: bug_rep.docx
  default-image-extension: png

metadata
  title: Reproducible Quarto Document
  jupyter: python3

Output created: bug_rep.docx

Reproducible Quarto Document
markdown

```
![截图 2024-06-21 14-45-28](https://github.com/quarto-dev/quarto-cli/assets/12797356/b018b7a7-2f02-4833-b4ce-939d9401f498)


quarto check result is :
```
quarto check
Quarto 1.5.45
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.45
      Path: /opt/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: v2024.06
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /home/jade/.TinyTeX/bin/x86_64-linux
      Version: 2024

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.6 (Conda)
      Path: /home/jade/anaconda3/envs/fnirs/bin/python
      Jupyter: 5.7.2
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.4.0
      Path: /usr/lib/R
      LibPaths:
        - /home/jade/R/x86_64-pc-linux-gnu-library/4.4
        - /usr/local/lib/R/site-library
        - /usr/lib/R/site-library
        - /usr/lib/R/library
      knitr: 1.46
      rmarkdown: 2.26

[✓] Checking Knitr engine render......OK

```

**Comment by hasibagen** [comment]
I also find a bug:
if the code is:

``````
---
title: "Reproducible Quarto Document"
format:  docx
params:
  datapath: null
execute:
  enable : true
  echo: false
  warning: false
  eval: true
jupyter: python3
---

```{python}
#| tags: [parameters]
#
```


```{python}
#| label: import
#| tags: [parameters]
print(datapath)
```
# markdown

``````

the result is

```
quarto render bug_rep.qmd -P datapath:1

Starting python3 kernel...Done

Executing 'bug_rep.quarto_ipynb'
  Cell 1/3: ''.........Done
  Cell 2/3: ''.........Done
  Cell 3/3: 'import'...Done

pandoc
  to: docx
  output-file: bug_rep.docx
  default-image-extension: png

metadata
  title: Reproducible Quarto Document
  jupyter: python3

Output created: bug_rep.docx
```

![image](https://github.com/quarto-dev/quarto-cli/assets/12797356/ee4751c1-83be-432c-a3e7-3300f8ec599c)


but if the code is

``````
---
title: "Reproducible Quarto Document"
format:  docx
params:
  datapath: null
execute:
  enable : true
  echo: false
  warning: false
  eval: true
jupyter: python3
---

```{python}
#| label: import
#| tags: [parameters]
print(datapath)
```

# markdown

``````
it will report:

```
quarto render bug_rep.qmd -P datapath:1

Executing 'bug_rep.quarto_ipynb'
  Cell 1/2: 'import'...ERROR:

An error occurred while executing the following cell:
------------------
print(datapath)
------------------


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 print(datapath)

NameError: name 'datapath' is not defined
```

**Reply by mcanouil** [reply]
Sorry but, did you look at the documentation regarding parameters and Jupyter?
From the example you shared, it does not seem to be the case.
- https://quarto.org/docs/computations/parameters.html#jupyter

`params` is for `knitr`, not Jupyter as described in the link above, thus the fact that it does not work.

**Reply by mcanouil** [reply]
Following the documentation:

````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
execute:
  enabled: true
---

```{python}
#| tags: [parameters]
datapath = None
```

```{python}
print(datapath)
```
````

````
quarto render index.qmd --execute-param "datapath:weird" --execute 
````

**Reply by hasibagen** [reply]
> Sorry but, did you look at the documentation regarding parameters and Jupyter? From the example you shared, it does not seem to be the case.
> 
> * https://quarto.org/docs/computations/parameters.html#jupyter
> 
> `params` is for `knitr`, not Jupyter as described in the link above, thus the fact that it does not work.

I read this document carefully, but maybe I didn't understand the meaning.

This document seems to say
For Jupyter, Quarto uses the same syntax for defining parameters as Papermill. To parameterize a document, designate a cell with the tag parameters and provide appropriate default values:

```{python}
#| tags: [parameters]

alpha = 0.1
ratio = 0.1
```

quarto render document.ipynb -P alpha:0.2 -P ratio:0.3

So I thought it was also possible to directly transfer parameters to ipynb. I tried
``````
---
title: "Reproducible Quarto Document"
format: docx
execute:
enable : true
echo: false
warning: false
eval: true
jupyter: python3
---
```{python}
#| label: import
#| tags: [parameters]
print(datapath)
```
quarto render document.ipynb -P datapath:0.1
``````
version

**Reply by hasibagen** [reply]
> ```
> ---
> title: "Reproducible Quarto Document"
> format: html
> engine: jupyter
> execute:
>   enabled: true
> ---
> 
> ```{python}
> #| tags: [parameters]
> datapath = None
> ```
> 
> ```{python}
> print(datapath)
> ```
> ```

Does this mean that if I want to use a parameter, I need to initialize it beforehand?
Because code like this will also result in an error.

``````
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
execute:
  enabled: true
---

```{python}
#| tags: [parameters]
print(datapath)
```

quarto render index.qmd --execute-param "datapath:weird" --execute

Executing 'index.quarto_ipynb'
  Cell 1/2: ''...ERROR:

An error occurred while executing the following cell:
------------------
print(datapath)
------------------


---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 print(datapath)

NameError: name 'datapath' is not defined
``````

**Reply by mcanouil** [reply, marked as answer]
You are still not following the documentation. `#| tags: [parameters]` is the code cell where you initialise the parameters, not use them.
Look at [my example](https://github.com/quarto-dev/quarto-cli/discussions/10081#discussioncomment-9836803) or the [documentation as a screenshot](https://quarto.org/docs/computations/parameters.html#jupyter) below:
<img width="883" alt="Screenshot of https://quarto.org/docs/computations/parameters.html#jupyter" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/021f0ff7-1460-4215-bc17-0c09ac881334">

**Reply by mcanouil** [reply]
As long as you don't follow the syntax/documentation, yes it will lead to error as expected.

**Reply by hasibagen** [reply]
> initialise the parameters

Now I understand what you mean, I understand that I need to initialise the parameters before using them.

**Reply by hasibagen** [reply]
> Following the documentation:
> 
> ```
> ---
> title: "Reproducible Quarto Document"
> format: html
> engine: jupyter
> execute:
>   enabled: true
> ---
> 
> ```{python}
> #| tags: [parameters]
> datapath = None
> ```
> 
> ```{python}
> print(datapath)
> ```
> ```
> 
> ```
> quarto render index.qmd --execute-param "datapath:weird" --execute 
> ```

I would like to ask if it is possible for you to also give a minimal demo of ipynb? Because I seem to have the same problem as in this post
https://github.com/quarto-dev/quarto-cli/discussions/9107

I have now fully implemented it using qmd, but I still have problems with ipynb.

Thank you very much for your help.

**Reply by mcanouil** [reply]
Use `quarto convert` on your file it should give you the correct Jupyter Notebook.

**Reply by hasibagen** [reply]
> Use `quarto convert` on your file it should give you the correct Jupyter Notebook.
Solved it, thank you very much for your help!

**Comment by hasibagen** [comment]
Another small problem that I encountered before is that it seems that the init cell will be automatically repeated once, and there cannot be a label name. If there is one, an error will be reported.

``````
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
execute:
  enabled: true
---

```{python}
#| label: prep
#| tags: [parameters]
datapath = None
```

```{python}
#| tags: [parameters]
print(datapath)
```
quarto render index.qmd -P datapath:4

Starting python3 kernel...Done

Executing 'index.quarto_ipynb'
  Cell 1/3: 'prep'...Done
  Cell 2/3: 'prep'...Done
  Cell 3/3: ''.......Done

ERROR: Cell label names must be unique (found duplicate '#prep')

Stack trace:
    at file:///opt/quarto/bin/quarto.js:30322:23
    at jupyterToMarkdown (file:///opt/quarto/bin/quarto.js:32297:9)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async Object.execute (file:///opt/quarto/bin/quarto.js:40543:25)
    at async renderExecute (file:///opt/quarto/bin/quarto.js:77997:27)
    at async renderFileInternal (file:///opt/quarto/bin/quarto.js:78165:43)
    at async renderFiles (file:///opt/quarto/bin/quarto.js:78033:17)
    at async render (file:///opt/quarto/bin/quarto.js:82893:21)
    at async Command.actionHandler (file:///opt/quarto/bin/quarto.js:83041:32)
    at async Command.execute (file:///opt/quarto/bin/quarto.js:8017:13)


``````

**Reply by mcanouil** [reply]
I'm pretty sure this has been previously reported but can't find the report right now.

**Reply by mcanouil** [reply]
Why do you have `#| tags: [parameters]` in a code cell that does not have parameters to be initialised?

**Reply by mcanouil** [reply]
FYI, I can reproduce on the latest development version (I'll open an issue, if I don't find the other one).

````qmd
---
format: html
engine: jupyter
---

```{python}
#| label: test
#| tags: [parameters]
datapath = None
```
````

### Trigger the crash by sending a new parameter value

```sh
quarto render index.qmd --execute-param datapath:"weird" --execute
```

```txt
Executing 'index.quarto_ipynb'
  Cell 1/2: 'test'...Done
  Cell 2/2: 'test'...Done

ERROR: Cell label names must be unique (found duplicate '#test')

Stack trace:
    at file:///Users/mcanouil/Projects/quarto/quarto-cli/src/core/jupyter/labels.ts:53:15
    at jupyterToMarkdown (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/core/jupyter/jupyter.ts:758:5)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async Object.execute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/jupyter/jupyter.ts:378:20)
    at async renderExecute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:232:25)
    at async renderFileInternal (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:592:35)
    at async renderFiles (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:325:9)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-shared.ts:102:18)
    at async Command.actionHandler (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/cmd.ts:248:26)
    at async Command.execute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/vendor/deno.land/x/cliffy@v1.0.0-rc.3/command/command.ts:1948:7)
```

### Keeping the default parameter value

```sh
quarto render index.qmd --execute
```

```txt
Executing 'index.quarto_ipynb'
  Cell 1/1: 'test'...Done

pandoc 
  to: html
  output-file: index.html
  standalone: true
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  
metadata
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  
Output created: index.html
```

**Reply by mcanouil** [reply]
@cscheid Do you recall the issue/discussion about the duplicated code cell? Here it is slightly different as this duplication associated with the label creates a crash.

**Reply by cscheid** [reply]
I don't. Please just create a new one.

**Reply by mcanouil** [reply]
- https://github.com/quarto-dev/quarto-cli/issues/10097
